### PR TITLE
add experimental_link_layers

### PR DESCRIPTION
**Pull request by tlambert03**
# Description
Here is an implementation of the idea that I proposed in https://github.com/napari/napari/issues/970#issuecomment-772813308 for linking certain characteristics of layers.  This is intended to complement, not replace, the concept of literal layer groups (#970).  

This will link certain (or all matching) attributes between two or more provided layers:

```python
from napari.experimental import link_layers, unlink_layers, layers_linked

viewer = napari.view_image(np.random.rand(3, 64, 64), channel_axis=0)

# link everything
link_layers(viewer.layers)

# link only certain attributes on some layers
link_layers(viewer.layers[:2], ('contrast_limits', 'gamma'))

# unlink one or more layers (with optional specific attributes)
unlink_layers(viewer.layers[0], ('opacity',))

# do it temporarily with a context manager:
with layers_linked(viewer.layers):
    ...
```

I tried to put it through it's paces and it seems to work very well.  But I figured it would be best to leave this one as "experimental" for now.

## Type of change

- [ ] Bug-fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# References

# How has this been tested?

- [x] the test suite for my feature covers most of the edge cases I could think of.  see tests.
- [x] all tests pass with my change

## Final checklist:
- [ ] My PR is the minimum possible work for the desired functionality
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works

### CODE REVIEW

**Comment by tlambert03** [comment]
@jwindhager, if you have a moment, I'd be curious for you to pull this PR and try it out for some of your layer-linking needs.  let me know if you have any questions on the API

**Comment by jwindhager** [comment]
Great! I will check asap. Was just about to submit a new feature request to untangle the discussion from #970 and will submit that one first, linking to this pull request afterwards if applicable.

**Comment by jwindhager** [comment]
Just gave this a try - this is incredible! Works very smoothly. As described in #2229, would be amazing if the linking would be interactive!

File: napari/utils/misc.py [hunk 1]
```diff
@@ -390,3 +398,56 @@ def ensure_list_of_layer_data_tuple(val):
         except TypeError:
             pass
     raise TypeError('Not a valid list of layer data tuples!')
+
+
+def get_equality_operator(obj: Any) -> Callable[[Any, Any], bool]:
```
Comment (sofroniewn) [inline napari/utils/misc.py]: Not needed in this PR, but I wonder if we'd start using this now in our EventedModel too. I know #2200 is still in progress, but might be nice to wire this up already.

File: napari/layers/utils/link_layers.py [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+
+        def _make_l2_setter(eq_op, l1=lay1, l2=lay2, attr=attr):
+            # where `eq_op` is an "equality checking function" that is suitable
+            # for the attribute object type.  (see ``get_equality_operator``)
+
+            def setter(event=None):
+                new_val = getattr(l1, attr)
+                # this line is the important part for avoiding recursion
+                if not eq_op(getattr(l2, attr), new_val):
+                    setattr(l2, attr, new_val)
```
Comment (sofroniewn) [inline napari/layers/utils/link_layers.py]: Once we have the EventedModel wired up which has it's own equality checks will we still need this?

File: napari/layers/utils/link_layers.py [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+#  in the form of (id(layer1), id(layer2), attribute_name)
+_LINKED: set[tuple[int, int, str]] = set()
+
+
+# attributes that make no sense to link, or error on changing
+# we may want to include colormap here, but in most cases, it works fine.
+PROHIBITED = {'thumbnail', 'status', 'name'}
+
+
+def _get_common_evented_attributes(
```
Comment (sofroniewn) [inline napari/layers/utils/link_layers.py]: Love this! Approach here is good. It might become even easier when we have the EventedModel, but what is here is good too

File: napari/layers/utils/link_layers.py [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+    from napari.layers import Layer
+
+#: Record of already linked layers... to avoid duplicating callbacks
+#  in the form of (id(layer1), id(layer2), attribute_name)
+_LINKED: set[tuple[int, int, str]] = set()
+
+
+# attributes that make no sense to link, or error on changing
+# we may want to include colormap here, but in most cases, it works fine.
+PROHIBITED = {'thumbnail', 'status', 'name'}
```
Comment (sofroniewn) [inline napari/layers/utils/link_layers.py]: I wonder if we should exclude `data` by default too?

File: napari/layers/utils/link_layers.py [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+
+    Not all layer events are attributes, and not all attributes have
+    corresponding events.  Here we get the set of valid, non-private attributes
+    that are both events and attributes for the provided layer set.
+
+    Parameters
+    ----------
+    layers : iterable
+        A set of layers to evaluate for attribute linking.
+    prohibited : set, optional
```
Comment (sofroniewn) [inline napari/layers/utils/link_layers.py]: Can we use `exclude` as a kwarg instead of `prohibited` - Pydantic uses exclude as part of its [serialization/ to dict API](https://pydantic-docs.helpmanual.io/usage/exporting_models/#modeldict) and I have a slight preference for it

File: napari/layers/utils/link_layers.py [hunk 6]
```diff
@@ -0,0 +1,145 @@
...
+           for attr in attributes:
+              lay1.events.<attr>.connect(_set_lay2_<attr>)
+
+    Recursion is prevented by checking for value equality prior to setting.
+
+    Parameters
+    ----------
+    layers : Iterable[napari.layers.Layer]
+        The set of layers to link
+    attributes : Iterable[str], optional
```
Comment (sofroniewn) [inline napari/layers/utils/link_layers.py]: I wonder maybe here if we can use `include` and `exclude` instead of `attributes` - thinking again about the pydantic [serialization/ to dict API](https://pydantic-docs.helpmanual.io/usage/exporting_models/#modeldict) which I quite like. If both `None` then get all common by default.

File: napari/layers/utils/link_layers.py [hunk 7]
```diff
@@ -0,0 +1,145 @@
...
+    # lastly, discard any method-only events (we just want attrs)
+    for attr in set(common_attrs):
+        # properties do not count as methods and will not be excluded
+        if ismethod(getattr(first_layer.__class__, attr, None)):
+            common.discard(attr)
+
+    return common
+
+
+def experimental_link_layers(
```
Comment (sofroniewn) [inline napari/layers/utils/link_layers.py]: Can you also provide an `experimental_unlink_layers`. One approach might be to let it take either a single layers iterable and unlink all the connection it can find between those layers or if two layers iterables are given then unlink all the connections between the first list and the second list. That could be quite nice if you say want to disconnect one layer from a group of others without disrupting connections within that group.

File: napari/layers/utils/link_layers.py [hunk 8]
```diff
@@ -0,0 +1,145 @@
...
+                f"layers: {extra}. Allowable attrs include:\n{valid_attrs}"
+            )
+    else:
+        # if no attributes are specified, ALL valid attributes are linked.
+        attr_set = valid_attrs
+
+    # now, connect requested attributes between all requested layers.
+    for (lay1, lay2), attr in product(permutations(layers, 2), attr_set):
+
+        key = (id(lay1), id(lay2), attr)
```
Comment (sofroniewn) [inline napari/layers/utils/link_layers.py]: Can you add a comment in the code here to the effect of
`# if the layers and attribute are already linked then ignore`

**Review by sofroniewn — COMMENTED** [review]
@tlambert03 I love this, I think it is going to be really nice a great way to solve a long standing feature request from @jni and @royerloic to allow for things like our contrast limits to be used to control multiple layers simultaneously.

I've left some comments on the PR most of which should be fairly simple to address, and I see ourselves able to merge this quite soon.

> would be amazing if the linking would be interactive!

absolutely @jwindhager! I'm actually now imagining a little chain-linked symbol that could appear in the layer item widget for all the linked layers like in photoshop - see

image

When you have multiple layers selected you can right-click on one of them and it brings up a menu where one of the options is to "link layers"

I think it is very nice that the individual layers themselves don't "know" who they are linked too (similarly one day our layers won't "know" they are selected), but I could also imagine an API where we add the concept of a set of keyword tags to a layer, so `layer.tags={'GFP', '5x', 'brain', `2021-02-06`}` for example and someone could then do `layers.link(tag='GFP')` and all layers which contained `GFP` would then get linked with each other. The could also be a simple way to view and link by tags in the GUI too. This would be nice as reader plugins could provide tags without worrying about linking.

Both of these ideas can come in follow-up PRs. For now it will just be great to get this in and start playing with it!!

File: napari/utils/misc.py (hunk 1 above)
Comment (tlambert03) [reply napari/utils/misc.py]: Yeah, I totally agree. As I think we discussed on zulip, it'd be nice if we can get away from `__equality_checks__` class attribute, and towards something that "just works" for the datatypes we use in napari.

File: napari/layers/utils/link_layers.py (hunk 2 above)
Comment (tlambert03) [reply napari/layers/utils/link_layers.py]: I _think_ maybe not?  In any case, whatever solution we come up with there can be directly applied here too, (whether that means that we use the same underling `eq_op` for both things, or that we just don't even need the check here, I'm not sure yet)

File: napari/layers/utils/link_layers.py (hunk 4 above)
Comment (tlambert03) [reply napari/layers/utils/link_layers.py]: yeah, that seems reasonable.

File: napari/layers/utils/link_layers.py (hunk 5 above)
Comment (tlambert03) [reply napari/layers/utils/link_layers.py]: yep!

File: napari/layers/utils/link_layers.py (hunk 6 above)
Comment (tlambert03) [reply napari/layers/utils/link_layers.py]: actually... for the function that the user calls, I'm not 100% sure.
calling `link_layers([layer1, layer2], attributes=('contrast_limits','gamma'))` reads (to me) clearer than `include=('contrast...')`.  The user may not know anything about the pydantic API and serialization, so I'm not sure how much that parity matters here?

File: napari/layers/utils/link_layers.py (hunk 2 above)
Comment (sofroniewn) [reply napari/layers/utils/link_layers.py]: Yeah that would be great! We'll still need it now for this PR as we're not using the eventedmodel on our layers yet, but I think it would be really nice to have an end state where we've taken care of this stuff there and so it's not needed, but we can see!!

File: napari/layers/utils/link_layers.py (hunk 7 above)
Comment (tlambert03) [reply napari/layers/utils/link_layers.py]: definitely, was just working on it! :)  will also add a context manager.

**Comment by tlambert03** [comment]
> > would be amazing if the linking would be interactive!

> absolutely @jwindhager! I'm actually now imagining a little chain-linked symbol that could appear in the layer item widget for all the linked layers like in photoshop - see

cool!  I think this should come in a followup PR, but this might be a perfect candidate for an item in a right-click context menu with a few layers selected?

**edit** *[rereads nicks comment and sees that he said exactly the same thing]*

File: napari/layers/utils/link_layers.py (hunk 6 above)
Comment (sofroniewn) [reply napari/layers/utils/link_layers.py]: do we want an `exclude` here? I think if we want both include and exclude then those words are nice. If we never want exclude then maybe attributes is fine. Let's get a third opinion from @jni. I would describe my preference here as weak, so happy with whatever

**Comment by sofroniewn** [comment]
> cool! I think this should come in a followup PR, but this might be a perfect candidate for an item in a right-click context menu with a few layers selected?

yes exactly, you could imaging options of `merge` layers, `split` layers or `link` layers. I was actually recently discussing something similar with @goanpeca @ppwadhwa @justinelarsen! But definitely this Qt stuff is for follow-up PRs

**edit** _[rereads talleys comment to see an edit saying exactly same thing - i guess we're both very excited about the idea]_

File: napari/layers/utils/link_layers.py (hunk 6 above)
Comment (tlambert03) [reply napari/layers/utils/link_layers.py]: I agree with that if we include an `exclude` argument.

**Comment by tlambert03** [comment]
question: do you prefer `from napari.layers.utils import experimental_link_layers`
or `from napari.experimental import link_layers`
(I kinda prefer the latter)

File: napari/layers/utils/link_layers.py (hunk 7 above)
Comment (tlambert03) [reply napari/layers/utils/link_layers.py]: ok, added `unlink_layers` and a `linked_layers` context manager.  i didn't do two lists of iterables: i did a single list.  If you provide a single layer, it will be unlinked from all other layers, if you provide a list, all combinations will be unlinked,  is that going to meet your imagined use case here?

**Comment by tlambert03** [comment]
ok, now we've got some nice link/unlink/context_manager action.  I think the main outstanding question here is how to name all of this.  I'd kinda like to make the module private `layers.utils._link_layers` and the function names short (without "experimental"), and expose it as `napari.experimental.link_layers` ... but let me know if you want to keep experimental directly in the function name ... in which case we could also do:
```python
# napari/experimental/__init__.py
from ..layers.utils._link_layers import experimental_link_layers as link_layers
```

File: napari/layers/utils/link_layers.py (hunk 7 above)
Comment (sofroniewn) [reply napari/layers/utils/link_layers.py]: yeah, i think that will work, i'm imagining the clicking behavior in the layer context menus and an option like `unlink layers` if you have one selected and `unlink layers` if you have multiple selected and I think that's the behavior i'd want to see

**Comment by sofroniewn** [comment]
I prefer `from napari.experimental import link_layers` too and keeping the name short. It's also much easier for people to then upgrade when something moves out of experimental as they just need to change the import and they don't need to change the function name. it will also be easier for us to have the function make that transition as we can just change the import too

File: napari/layers/utils/link_layers.py (hunk 7 above)
Comment (tlambert03) [reply napari/layers/utils/link_layers.py]: yeah, I think this should work then.  basically, if you click on the "lock" icon on any individual layer, it will call `unlink_layer([layer])` and "free" it from connections

**Comment by tlambert03** [comment]
great!

File: napari/layers/utils/_link_layers.py [hunk 9]
```diff
@@ -0,0 +1,203 @@
...
+    """Unlink previously linked ``attributes`` between all layers in ``layers``.
+
+    Parameters
+    ----------
+    layers : Iterable[napari.layers.Layer]
+        The list of layers to unlink.  All combinations of layers provided will
+        be unlinked.  If a single layer is provided, it will be unlinked from
+        all other layers.
+    attributes : Iterable[str], optional
+        The set of attributes to link.  If not provided, all connections
```
Comment (sofroniewn) [inline napari/layers/utils/_link_layers.py]:
```suggestion
        The set of attributes to unlink.  If not provided, all connections
```

File: napari/layers/utils/_link_layers.py [hunk 10]
```diff
@@ -0,0 +1,203 @@
...
+    else:
+        layer_combos = {frozenset(i) for i in combinations(layer_ids, 2)}
+        keys = (k for k in list(_LINKED_LAYERS) if set(k[:2]) in layer_combos)
+    if attributes:
+        keys = (k for k in keys if k[2] in attributes)
+    _unlink_keys(keys)
+
+
+@contextmanager
+def linked_layers(layers: Iterable['Layer'], attributes: Iterable[str] = ()):
```
Comment (sofroniewn) [inline napari/layers/utils/_link_layers.py]: hmm i wonder about this name - when I first saw it I thought of the `_LINKED_LAYERS`. @tlambert03 in zulip just suggested `layers_linked`

File: napari/layers/utils/_tests/test_link_layers.py [hunk 11]
```diff
@@ -0,0 +1,156 @@
...
+    l2 = layers.Points(None)
+    l3 = layers.Points(None)
+    link_layers([l1, l2, l3])
+
+    l1.opacity = 0.5
+    assert l1.opacity == l2.opacity == l3.opacity == 0.5
+
+    # if we delete layer3 we shouldn't get an error when updating otherlayers
+    del l3
+    l1.opacity = 0.25
```
Comment (sofroniewn) [inline napari/layers/utils/_tests/test_link_layers.py:95]: Did this work because we sorted out our disconnecting events on deletion stuff, or did you have to do something special?

File: napari/layers/utils/_tests/test_link_layers.py [hunk 12]
```diff
@@ -0,0 +1,156 @@
...
+    l3 = layers.Points(None)
+    assert len(l1.events.opacity.callbacks) == 0
+    with linked_layers([l1, l2, l3], ('opacity',)):
+        assert len(l1.events.opacity.callbacks) == 2
+        assert len(l1.events.blending.callbacks) == 0  # it's just opacity
+        del l2  # if we lose a layer in the meantime it should be ok
+    assert len(l1.events.opacity.callbacks) == 0
+
+
+def test_unlink_layers():
```
Comment (sofroniewn) [inline napari/layers/utils/_tests/test_link_layers.py:112]: Love this!

**Review by sofroniewn — APPROVED** [review]
Ok, all done - beautiful work @tlambert03, I'm excited to get this merged. I'll let @jni weigh in if he has any additional comments.

Maybe @jwindhager can keep on testing this in the wild, it would be great to know if you're able to incorporate it into your work and let us know how it's going!!

File: napari/layers/utils/link_layers.py (hunk 6 above)
Comment (sofroniewn) [reply napari/layers/utils/link_layers.py]: let's leave with attributes, i don't think `exclude` is critical

File: napari/layers/utils/_tests/test_link_layers.py (hunk 11 above)
Comment (tlambert03) [reply napari/layers/utils/_tests/test_link_layers.py:95]: aha!  I didn't do anything special,  I wanted to test for it though and it just worked.  that must be it!

Comment (tlambert03) [reply napari/layers/utils/_tests/test_link_layers.py:95]: or not :) but i started to implement `weakref` stuff and then just checked it if worked without, and it did

Comment (sofroniewn) [reply napari/layers/utils/_tests/test_link_layers.py:95]: great test! i'm glad to see it, connections hanging around etc was one thing i was maybe worried about. We can continue to monitor as we use this more in the wild too

**Comment by jwindhager** [comment]
> absolutely @jwindhager! I'm actually now imagining a little chain-linked symbol that could appear in the layer item widget for all the linked layers like in photoshop - see
> 
> image

Love the idea, would be so useful! I just see issues arising re multiple link groups (if supported?) and linking layers that are already linked to other layers. How is this solved in photoshop? I also like the fact that there is a Channels tab :slightly_smiling_face: #2229

> Maybe @jwindhager can keep on testing this in the wild, it would be great to know if you're able to incorporate it into your work and let us know how it's going!!

Happy to play around in a bit more detail with this! However, as far as the use case of napari-imc goes, this is somewhat different, as I need to explicitly provide the user an interface to the list of channels anyway, where each channel item internally holds a list of layers associated with this channel and can therefore directly propagate to all channels already. But maybe it could replace the redundant channel controls widget in napari-imc - I just fear that in the case of napari-imc, the list of layers is too extensive for the user to navigate directly (compared to the list of channels). In any case, I'll have a look!

File: napari/layers/utils/link_layers.py (hunk 6 above)
Comment (jni) [reply napari/layers/utils/link_layers.py]: I prefer attributes — I don't see much use for exclude here.

**Comment by sofroniewn** [comment]
@tlambert03 can you update this with the changes from #2232. I think this will then be good to merge!!

**Review by jni — APPROVED** [review]
@sofroniewn I'll let you do the merging since I've made a few minor changes since your last review!

**Comment by sofroniewn** [comment]
Ok this looks great! thanks for that review @jni - i see @tlambert03 has done a final pass and clean up too, so I think this is good to go! Will merge now!!
